_posts/2024-06-26-benchmarks-over-time.md

```diff
@@ -0,0 +1,608 @@
+
+#### Summary Metrics
+
+With the exception of the scale tests, each benchmark was run 3 times and the median time was used for reporting results.
+The scale tests were run once and produced a binary metric, success or failure, at each data size tested.
+As older versions would not fail gracefully, the scale metrics were accumulated across multiple partial runs.
+
+#### Computing Resources
+
+All tests use a Macbook Pro M1 with 16GB of RAM.
+In 2024, this is far from state of the art! 
+If you have more powerful hardware, you will see both improved performance and scalability.
+
+#### DuckDB Versions
+
+Version 0.2.7, published in June 2021, was the first version to include a Python client compiled for ARM64, so it was the first version that could easily run on the benchmarking compute resources.
+Version 1.0.0 is the latest available at the time of publication (June 2024), although we also provide a sneak preview of an in-development feature branch.
+
+#### Default Settings
+
+All versions were run with the default settings.
+As a result, improvements from a new feature only appear in these results once that feature became the default and was therefore ready for production workloads.
+
+
+### Window Functions Benchmark
+
+Each benchmark query follows the format below, but with different sets of window functions in the `⟨window function(s)⟩` placeholder.
+The table in use is the largest table from the H2O.ai join benchmark, and in this case the 5GB scale was used.
+
+```sql
+DROP TABLE IF EXISTS windowing_results;
+
+CREATE TABLE windowing_results AS
+    SELECT 
+        id1,
+        id2,
+        id3,
+        v2,
+        ⟨window function(s)⟩
+    FROM join_benchmark_largest_table;
+```
+
+The various window functions that replace the placeholder are below and are labelled to match the result graphs.
+These were selected to showcase the variety of use cases for window functions, as well as the variety of algorithms required to support the full range of the syntax.
+The DuckDB documentation contains a [full railroad diagram of the available syntax]({% link docs/sql/window_functions.md %}#syntax).
+If there are common use cases for window functions that are not well-covered in this benchmark, please let us know!
+
+```sql
+/* 302 Basic Window */ 
+sum(v2) OVER () AS window_basic
+
+/* 303 Sorted Window */
+first(v2) OVER (ORDER BY id3) AS first_order_by,
+row_number() OVER (ORDER BY id3) AS row_number_order_by
+
+/* 304 Quantiles Entire Dataset */
+quantile_cont(v2, [0, 0.25, 0.50, 0.75, 1]) OVER ()
+    AS quantile_entire_dataset
+
+/* 305 PARTITION BY */
+sum(v2) OVER (PARTITION BY id1) AS sum_by_id1,
+sum(v2) OVER (PARTITION BY id2) AS sum_by_id2,
+sum(v2) OVER (PARTITION BY id3) AS sum_by_id3
+
+/* 306 PARTITION BY ORDER BY */
+first(v2) OVER
+    (PARTITION BY id2 ORDER BY id3) AS first_by_id2_ordered_by_id3
+
+/* 307 Lead and Lag */
+first(v2) OVER
```

Comment:
are these window function calls `first(v2)` instead of `lag(v2)` and `lead(v2)?

Reply:
Yes, just because I like to use the generic window function syntax since I have it memorized by now. Functionally the same though, at least that is the intention!